compiler/rename/RnExpr.hs

```diff
@@ -324,10 +324,22 @@ rnExpr (RecordUpd { rupd_expr = expr, rupd_flds = Right flds})
               addErr $ text "For this to work enable NamedFieldPuns."
          ; (getField, fv_getField) <- lookupSyntaxNameLocalEnv (mkVarOcc "getField")
          ; (setField, fv_setField) <- lookupSyntaxNameLocalEnv (mkVarOcc "setField")
-         ; (e, fv_e) <- rnLExpr expr
+         -- We preceed the record dot update with a
+         --
+         -- 'case record_expr of
+         --    x -> setField ... x'
+         --
+         -- for better type inference.
+         ; let arg_lit = fsLit "arg"
+               arg_name = mkSystemVarName (mkVarOccUnique arg_lit) arg_lit
+               arg_syn_expr = mkRnSyntaxExpr arg_name
+               arg_pat = noLoc $ VarPat noExt $ noLoc arg_name
+               arg_var = syn_expr arg_syn_expr
+         ; (expr', fv_expr) <- rnLExpr expr
          ; (us, fv_us) <- rnHsUpdProjs flds
-         ; return ( mkRecordDotUpd getField setField e us
-                  , plusFVs [fv_getField, fv_setField, fv_e, fv_us] )
+         ; let match = mkHsCaseAlt arg_pat (noLoc $ mkRecordDotUpd getField setField (noLoc arg_var) us)
```

Comment:
let me comment here instead of on the other PR, I think that’s easier to understand: Can we drop the `case` if `us` is a singleton list?

Reply:
eh sorry I mean that the field names in `us` are singleton lists.

Reply:
I think `us` is the updates, not the fields. So to check whether it is a nested update, you have dig deeper and check each update.

Reply:
Right that’s what I realized in my correction. Doesn’t seem that hard though I hope? (famous last words)

Reply:
No, should be possible. I'll check it out.